packages/@aws-cdk/aws-servicecatalog/README.md

```diff
@@ -184,6 +184,48 @@ const product = new servicecatalog.CloudFormationProduct(this, 'Product', {
 });
 ```
 
+We can also define a `versioningStrategy` for your `ProductStack` deployment. 
+With the `Default` strategy, a new `ProductStack` is needed for each `productVersion`
+and each `productVersion` will get overwritten with the latest changes to your `ProductStack`.
+With the `RetainPreviousVersions` strategy, a local template will be retained for each `ProductStack` deployed.
+Subsequently, any `ProductStack` changes will not overwrite the current version.
+A new `productVersionName` must be specified in order for `ProductStack` changes to be deployed.
+When the `ProductStack` is updated with a new version, the previous version can still be deployed.
+The template of previously deployed `ProductStack` can be referenced using `fromProductStackContext`
+and passing the corresponding `ProductStack` id.
+
+After using the `RetainPreviousVersions` strategy to deploy version `v1` of your `ProductStack`,
+the `ProductStack` is updated to version `v2` and version `v1` can be deployed from context:
+
+```ts
+import * as s3 from '@aws-cdk/aws-s3';
+import * as cdk from '@aws-cdk/core';
+
+class S3BucketProduct extends servicecatalog.ProductStack {
+  constructor(scope: cdk.Construct, id: string) {
+    super(scope, id);
+
+    new s3.Bucket(this, 'BucketProductV2');
+  }
+}
+
+const product = new servicecatalog.CloudFormationProduct(this, 'MyFirstProduct', {
+  productName: "My Product",
+  owner: "Product Owner",
+  productVersions: [
+    {
+      productVersionName: "v2",
+      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(new S3BucketProduct(this, 'S3BucketProduct')),
+      versioningStrategy: servicecatalog.VersioningStrategy.RETAIN_PREVIOUS_VERSIONS
```

Comment:
`versioningStrategy` only applies to `fromProductStack` looks like... so shouldn't it be part of the `fromProductStack` constructor?

Reply:
@rix0rrr is correct. The `versioningStrategy` currently only applies to ProductStacks.
The reason we cannot set it directly on the product is because we may have two separate productVersions for a single product.
For example a product may have:
`v1_blue` and `v1_green`, which can both be updated to a `v2` version. We could have a strategy for the entire product, but that will enforce that strategy for each version. Making it tied to each version allows the user to have one use `DEFAULT` and another use `RETAIN_PREVIOUS_VERSIONS`

Setting a versioningStrategy on `v1` would not do anything since it using `fromProductStackContext`. Think of the versioningStrategy as applying to that specific ProductStack rather than to a specific version.

The reason we choose not to make it apart of the `fromProductStack` constructor is because we want it leave it open to have future versioningStrategies that don't apply to just `ProductStack`. For example, we could add the `RETAIN_PREVIOUS_VERSIONS` or another strategy to `fromAsset` if there is a useful use case for it. However, I did originally have it in the constructor, but was suggested by Adam (left CDK team) to move it. 
I can see how it is confusing, so if you think we should have it tied only to the `fromProductStack` constructor we can change this and we should still be able to make it more flexible in the future by adding an optional field to the other constructors.

Reply:
I'm not sure I agree `versioningStrategy` applies only to `fromProductStack()`. A Product Version from an Asset (created using `fromAsset()`), or referencing some file  uploaded somewhere (created using `fromUrl()`)  can be changed in the same way one created from `ProductStack` can be, and so you might want to prevent those changes from being deployed the same way you want to prevent changes to `ProductStack` from being deployed.

Reply:
Right we could extend the same functionality to `fromAsset()` or `fromUrl()`, however this PR itself is focusing on ProductStack since that was the issue a customer brought up with us.
With the other two the user can version their templates themselves either with additoinal local asset or from ie. git (and they pass in the additional urls). 
The only way to version `fromProductStack()` is to create an entirely new ProductStack in code, which doesn't make sense the more versions we have and could result in a lot of duplicate code. We are trying to make it easier for them to do this by automatically saving their generated templates instead and making it easy to reference those templates.

Reply:
That doesn't really make any sense though. You're allowing this to be set on the Product Version level, so this should work with all CloudFormation Template variants.

If you're only going to make this work for `fromProductStack()`, then Rico's suggestion is definitely right, and you should move this property into the the `fromProductStack()` method, and remove it from the Product Version.

Reply:
> Right we could extend the same functionality to `fromAsset()` or `fromUrl()`, however this PR itself is focusing on ProductStack since that was the issue a customer brought up with us. With the other two the user can version their templates themselves either with additoinal local asset or from ie. git (and they pass in the additional urls). The only way to version `fromProductStack()` is to create an entirely new ProductStack in code, which doesn't make sense the more versions we have and could result in a lot of duplicate code. We are trying to make it easier for them to do this by automatically saving their generated templates instead and making it easy to reference those templates.

Reply:
It can be made to work with any CloudFormation Template variants, it is just not implemented in this PR.
Since ProductStack has its own class where we can implement the retain logic in the class itself, while the other two havevery simple logic that will just pass the templateUrl to the Product, we would have to introduce a completely independent class to retain and apply it to all those. 
At this point, I'm not sure if its worth the effort to do that especially since they have other means to easily manually version in those variants, so it might not be as useful.

Reply:
In my opinion, it's clearly worth it, because:

1. You want to prevent updating existing versions that use templates other than `fromProductStack()` as well (see here: https://github.com/aws/aws-cdk/pull/20244#discussion_r872633710 on my reasoning why is this useful).
2. You're introducing a very confusing API (you're putting a property on Product Version that only works with one CloudFormation Template variant) to a stable module.

Reply:
`The issue I'm talking about here is not additional versioning. The problem is overriding an existing version by changing a file (either on disk with fromAsset(), or somewhere most likely in S3 with fromUrl()). Pretty sure you want to prevent that from happening, for the same reasons you want to prevent that from happening for fromProductStack().`

Agreed,  try to prevent overwrites from any  CloudFormation Template variants happening could be useful. Do you think we should redesign this PR to have those additional features, or build on that feature with an additional PR.

`Why is it OK to have multiple files for fromAsset() or fromUrl(), but it's a problem for fromProductStack()? I don't see why copy&pasting a JSON/YAML file is fine, but having multiple version of a class is a problem?`

Not saying its okay. It's just have a multiple versions of a class is a bit harder to maintain. They would need to create a new file for each ProductStack class.

`That's not true though: ...`
The ProductStack may have more complicated logic since a ProductStack can contain any number of other AWS resources.

Reply:
In my opinion, the PR cannot be merged in its current form. Given that, you have two choices:

1. Keep the "template is saved in a file" functionality only to `fromProductStack()`. But in that case, you need to remove `versioningStrategy` from `ProductVersion`, and move it to `fromProductStack()`.
2. Keep the `versioningStrategy` in `ProductVersion`, but then you need to make it work with all variants of `CloudFormationTemplate`, not only with `fromProductStack()`.

I'll let you decide which one do you prefer. @rix0rrr and @comcalvi might also have opinions on which one they prefer.

Reply:
Does it make sense to extend it to `fromUrl()` when this is an external URL. 
With `fromAsset()` we would just create a copy of the asset file. 
However for `fromUrl()` the only way it would makes sense is to read in the file from the URL and make a local copy. Which I'm not sure it is something we want to do.
This makes me lean more towards making `versioningStrategy` tied to `fromProductStack()`

Reply:
That depends on the behavior of the ServiceCatalog service.

When updating a Product, does it re-read the file from the URL given?

If it doesn't, you can just save the URL itself locally (instead of the entire contents of the template) - it will do what you want.

If it does re-read the template from the URL, then yes, you need to download it at synth time, and save it locally.

Reply:
I believe it does re-read the template from the URL, so I would have to download it and save it locally.
However, saving the contents of the template locally, feels like it defeats the purpose of them storing it externally and then passing an URL. Although it would be up to them to specify they want this

Reply:
I guess you can also just store only the URL anyway, and simply assume that, as long as the URL stays the same, the versions are the same from CDK's perspective.

That means a customer can theoretically still change them by switching the file under the URL, but maybe that's acceptable.

Reply:
Another argument against that would be that the customer now would have to specify what the source TemplateType is.
 
Since storing the URL would take different behavior, as we have to pass the URL directly to the Product instead of creating an Asset from the file and passing the URL of the Asset. We need to know which one to perform. 

Meaning that if a customer want to retrieve a previous version, they need to know what the original TemplateType was, which doesn't make sense at all. Unless we want to implement logic to infer the type based on the contents of the file.

There is also the issue of checking if a product version is already existing, which we might need additional logic as well since we were originally using hash of the content of the file to verify.

Reply:
Not sure I understand. If you store the URL, you would also pass the URL for the Version for that Product Version.

Reply:
I mean when we try to read it from our local context. The user would only be passing in the productVersionName they would like to retrieve. We would need to determine if this productVersionName was stored from a productStack or from a URL (url would be stored in file instead of contents) as they would take different logic.

Reply:
No one is asking us to retain Assets or retain URLs, and it seems like its will add more complexity, edge cases and possibly unintended behavior for the user.

I think its better to focus on the customer ask here and only retain Product Stacks. Change`VersioningStrategy` to `RetentionStrategy` as suggested by Adam and scope it down to a field in `fromProductStack()`

Reply:
> I mean when we try to read it from our local context. The user would only be passing in the productVersionName they would like to retrieve. We would need to determine if this productVersionName was stored from a productStack or from a URL (url would be stored in file instead of actual contents) as they would take different logic.

Of course, but that should be easy, right? We are the ones who control the format of these files, and the logic that both writes, and reads them. So I don't think there should be any issues with implementing this.